File: Exercises/rb101-rb109_small_problems/medium_2/lettercase_percentage.rb
=begin
# Problem
- input: string
- output: hash
- rules:
  - input can contain lowercase and uppercase letters, spaces, numbers, non-letter characters
  - spaces count as 'neither'
  - output is a hash with three keys:
    - lowercase percentage
    - uppercase percentage
    - neither percentage
  - output values are floats

# Algorithm
- Initialize a local variable `results` as a hash with default value 0
- Split input string into an array of characters with parameter `char` and iterate over the characters:
  - results[:lowercase] += 1 if `char` is lowercase
  - results[:uppercase] += 1 if `char` is uppercase
  - results[:neither] += 1 if `char` is neither
- Convert values to percentages
- Return `results`

=end

def letter_percentages(string)
  results = {}
  characters = string.chars

  results[:lowercase] = characters.count { |char| char =~ /[a-z]/ }
  results[:uppercase] = characters.count { |char| char =~ /[A-Z]/ }
  results[:neither] = characters.count { |char| char =~ /[^a-zA-Z]/ }

  results[:lowercase] = ((results[:lowercase] / string.size.to_f) * 100).round(2)
  results[:uppercase] = ((results[:uppercase] / string.size.to_f) * 100).round(2)
  results[:neither] = ((results[:neither] / string.size.to_f) * 100).round(2)

  results
end

p letter_percentages('abCdef 123') #== { lowercase: 50.0, uppercase: 10.0, neither: 40.0 }
p letter_percentages('AbCd +Ef') #== { lowercase: 37.5, uppercase: 37.5, neither: 25.0 }
p letter_percentages('123') #== { lowercase: 0.0, uppercase: 0.0, neither: 100.0 }
p letter_percentages('abcdefGHI')

File: ls215/exercises/small_problems/medium_2/lettercase_percentage.js
/* eslint-disable */

/*

### Problem Description ###

Write a function that takes a string and returns an object containing the following three properties:

- the percentage of characters in the string that are lowercase letters
- the percentage of characters that are uppercase letters
- the percentage of characters that are neither

You may assume that the string will always contain at least one character.

# Problem
- input: string
- output: object (with 3 properties)
- rules:
  - explicit:
    - given a string as input, return an object with three properties:
      1. percentage of lowercase characters
      2. percentage of uppercase characters
      3. percentage of non-letter characters
    - object values are percentages, with two decimal places, as strings
    - string will always contain at least 1 character
  - implicit:
    - spaces count as non-letter characters

# Data Structures
- input: string
- working with input:
  - string
  - numbers (for percentages)
- output: object

# Algorithm
Approach:
  - create an object with three properties: lowercase, uppercase, neither, with initial values of 0.
  - iterate over the input string:
    - for each character, incremenet the appropriate object property count depending on the type of character: lowercase, uppercase, neither
  - once all characters have been counted, iterate over the object properties, converting each property's value to a percentage as a string
  - return the object

*/

/* eslint-enable */

function letterPercentages(string) {
  let percentages = { lowercase: 0, uppercase: 0, neither: 0 };

  letterCount(string, percentages);

  Object.keys(percentages).forEach(key => {
    percentages[key] = String(
      ((percentages[key] / string.length) * 100).toFixed(2)
    );
  });

  return console.log(percentages);
}

function letterCount(string, percentagesObject) {
  const lowercase = /[a-z]/;
  const uppercase = /[A-Z]/;

  for (let index = 0; index < string.length; index += 1) {
    let char = string[index];

    switch (true) {
      case lowercase.test(char):
        percentagesObject.lowercase += 1;
        break;
      case uppercase.test(char):
        percentagesObject.uppercase += 1;
        break;
      default:
        percentagesObject.neither += 1;
        break;
    }
  }
}

letterPercentages('abCdef 123');
// { lowercase: "50.00", uppercase: "10.00", neither: "40.00" }

letterPercentages('AbCd +Ef');
// { lowercase: "37.50", uppercase: "37.50", neither: "25.00" }

letterPercentages('123');
// { lowercase: "0.00", uppercase: "0.00", neither: "100.00" }
